Upgrade services from iceberg 1.2.0 to iceberg 1.5.2

jiang95-dev opened the pull request:
## Summary
Upgrade `services` from iceberg 1.2.0 to iceberg 1.5.2. `Integrations`, `apps`, and `tables-test-fixtures` will remain iceberg 1.2.0.

Notable changes from iceberg 1.2.0 to iceberg 1.5.2 are not limited as following:
- Add view support
- Add FileIO that supports ADLSv2 storage
- Support file and partition delete granularity
- Track partition statistics in TableMetadata
- Add last updated timestamp and snapshot ID to partitions metadata table
- Added support for Spark 3.5 and removed support for Spark 3.2
- Add fast_forward procedure

## Changes
Added `openhouse.iceberg-conventions-1.2` to strictly pin the dependencies version to be 1.2.0. 

Excluded the iceberg libraries from the `tables` and `tables-test-fixtures` module to remove the circular dependency.

**After the change, modules using iceberg 1.5.2 will be**:
- services
- cluster:storage
- internalcatalog
- htscatalog

**Modules using iceberg 1.2.0 will be**:
- apps
- integrations
- libs
- tables-test-fixtures

**Checkboxes**:
- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [ ] Bug Fixes
- [ ] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [x] Refactoring
- [ ] Documentation
- [ ] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done
Comprehensive spark compatibility tests performed in the local cluster: https://docs.google.com/document/d/1yXORH5ety5Gdr6Avsh7XEmIghiJ8wQrg0l4G4_9Soxo/edit#heading=h.h69v5xcbp1md

- [ ] Manually Tested on local docker setup. Please include commands ran, and their output.
- [ ] Added new tests for the changes made.
- [x] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

sumedhsakdeo commented on buildSrc/src/main/groovy/openhouse.iceberg-conventions-1.2.gradle:
```diff
@@ -0,0 +1,11 @@
+}
+
+dependencies {
+  implementation('org.apache.iceberg:iceberg-bundled-guava:' + icebergVersion + "!!")
```
Curious what !! do?

jiang95-dev replied:
It forcefully pins the version when there's a conflict.

sumedhsakdeo commented on buildSrc/src/main/groovy/openhouse.iceberg-conventions.gradle:
```diff
@@ -1,5 +1,5 @@
 ext {
-  icebergVersion = '1.2.0'
+  icebergVersion = '1.5.2'
```
Why is name not iceberg-conventions-1.5.2

jiang95-dev replied:
Yeah we can change that name too.

sumedhsakdeo replied:
Cool let's do it

sumedhsakdeo commented on libs/datalayout/build.gradle:
```diff
@@ -1,7 +1,6 @@
 plugins {
     id 'openhouse.java-conventions'
     id 'openhouse.hadoop-conventions'
-    id 'openhouse.iceberg-conventions'
```
Why?

jiang95-dev replied:
Added back.

HotSushi reviewed (changes requested):
Might be looked into already, but is there any spec changes between 1.2 to 1.5? also can we do a staging test before merge?

jiang95-dev commented:
> Might be looked into already, but is there any spec changes between 1.2 to 1.5? also can we do a staging test before merge?

Added some notable changes, but there's no changes in spec.
Also added the spark compatibility testing report.

autumnust commented on iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java:
```diff
@@ -106,8 +106,9 @@ void testDoCommitAppendSnapshotsInitialVersion() throws IOException {
       Assertions.assertEquals(
-          4,
-          updatedProperties.size()); /*location, lastModifiedTime, version and appended_snapshots*/
+          5,
```
`write.parquet.compression-codec` why is this showing up as part of iceberg version upgrade?

jiang95-dev replied:
It is explicitly set by iceberg in 1.5.2: https://github.com/apache/iceberg/blob/5e139f48505b9b622e2cf389eae2bb74d6420510/core/src/main/java/org/apache/iceberg/TableMetadata.java#L93

autumnust reviewed (changes requested):
one question. 

Also, when we have things like ADLSv2 storage, can we remove Kai's copy ?

jiang95-dev commented:
> one question.
> 
> Also, when we have things like ADLSv2 storage, can we remove Kai's copy ?

I want to do that in a separate PR.

abhisheknath2011 commented on apps/spark/build.gradle:
```diff
@@ -1,7 +1,7 @@
     id 'openhouse.java-conventions'
     id 'openhouse.hadoop-conventions'
-    id 'openhouse.iceberg-conventions'
+    id 'openhouse.iceberg-conventions-1.2'
```
Why this module is using iceberg 1.2?

jiang95-dev replied:
We are only upgrading the server for now.

abhisheknath2011 commented on buildSrc/src/main/groovy/openhouse.iceberg-conventions-1.2.gradle:
```diff
@@ -0,0 +1,11 @@
+ext {
+  icebergVersion = '1.2.0'
```
Is this convention for backward compatibility reason?

jiang95-dev replied:
The plan is upgrade the server and client separately.

abhisheknath2011 commented on tables-test-fixtures/build.gradle:
```diff
@@ -2,6 +2,7 @@ plugins {
   id 'java'
   id 'com.github.johnrengelman.shadow' version '7.1.2'
   id 'openhouse.maven-publish'
+  id 'openhouse.iceberg-conventions-1.2'
```
Why some modules are on Iceberg 1.2?

jiang95-dev commented:
Performance the same [tests](https://docs.google.com/document/d/1yXORH5ety5Gdr6Avsh7XEmIghiJ8wQrg0l4G4_9Soxo/edit) in the staging environment. @HotSushi
